src/qs8-gemm/MRx4c16-wasmusdot.c.in

```diff
@@ -0,0 +1,230 @@
+      if XNN_UNPREDICTABLE(mr < ${M+1}) {
+        a${M} = a${M-1};
+        c${M} = c${M-1};
+      }
+
+  const v128_t vsign_mask = wasm_u8x16_const_splat(UINT8_C(0x80));
+  do {
+    $if DATATYPE == "QD8":
+      v128_t vksum0 = wasm_v128_load32_zero(w);
+      $for N in range(1, 4):
+        v128_t vksum${N} = wasm_v128_load32_zero((const int32_t*) w + ${N});
+      $for M in range(MR):
+        const v128_t vinput_zero_point${M} = wasm_v128_load32_splat(&quantization_params[${M}].zero_point);
+      $for M in range(MR):
+        $for N in range(4):
+          v128_t vacc${M}x${N} = wasm_i32x4_mul(vksum${N}, vinput_zero_point${M});
+    $else:
+      v128_t vacc0x0 = wasm_v128_load32_zero(w);
+      $for N in range(1, 4):
+        v128_t vacc0x${N} = wasm_v128_load32_zero((const int32_t*) w + ${N});
+      $for M in range(1, MR):
+        $for N in range(4):
+          v128_t vacc${M}x${N} = vacc0x${N};
+    w = (const int32_t*) w + 4;
+
+    size_t k = kc;
+    do {
+      $for M in range(MR):
+        const v128_t va${M} = wasm_v128_xor(wasm_v128_load(a${M}), vsign_mask);
+        a${M} += 16;
+
+      $for N in range(4):
+        $if N == 0:
+          const v128_t vb${N} = wasm_v128_load(w);
+        $else:
+          const v128_t vb${N} = wasm_v128_load((const int8_t*) w + ${N * 16});
+
+        $for M in range(MR):
+          vacc${M}x${N} = wasm_i32x4_relaxed_dot_i8x16_i7x16_add(vb${N}, va${M}, vacc${M}x${N});
+
+      w = (const int8_t*) w + 64;
+      k -= 16 * sizeof(int8_t);
+    } while (k != 0);
+
+    $for M in range(MR):
+      const v128_t vacc${M}x02 = wasm_i32x4_add(wasm_v32x4_shuffle(vacc${M}x0, vacc${M}x2, 0, 4, 1, 5), wasm_v32x4_shuffle(vacc${M}x0, vacc${M}x2, 2, 6, 3, 7));
+      const v128_t vacc${M}x13 = wasm_i32x4_add(wasm_v32x4_shuffle(vacc${M}x1, vacc${M}x3, 0, 4, 1, 5), wasm_v32x4_shuffle(vacc${M}x1, vacc${M}x3, 2, 6, 3, 7));
+
+    $for M in range(MR):
+      v128_t vacc${M}x0123 = wasm_i32x4_add(wasm_v32x4_shuffle(vacc${M}x02, vacc${M}x13, 0, 4, 1, 5), wasm_v32x4_shuffle(vacc${M}x02, vacc${M}x13, 2, 6, 3, 7));
+
+    $for M in range(MR):
+      vacc${M}x0123 = wasm_f32x4_convert_i32x4(vacc${M}x0123);
+
+    $if DATATYPE == "QD8":
+      $for M in range(MR):
+        const v128_t vinput_scale${M} = wasm_v128_load32_splat(&quantization_params[${M}].inv_scale);
+
+      $for M in range(MR):
+        vacc${M}x0123 = wasm_f32x4_mul(vacc${M}x0123, vinput_scale${M});
+
+      const v128_t vfilter_output_scale0123 = wasm_v128_load(w);
+      $for M in range(MR):
+        vacc${M}x0123 = wasm_f32x4_mul(vacc${M}x0123, vfilter_output_scale0123);
+      w = (const float*) w + 4;
+
+      const v128_t vbias0123 = wasm_v128_load(w);
+      $for M in range(MR):
+        vacc${M}x0123 = wasm_f32x4_add(vacc${M}x0123, vbias0123);
+      w = (const float*) w + 4;
+
+      const v128_t vmin = wasm_v128_load64_splat(params->wasmsimd.min);
+      $for M in range(MR):
+        vacc${M}x0123 = wasm_f32x4_pmax(vacc${M}x0123, vmin);
```

Comment:
note pmax is 2 instructions on arm.

Reply:
(I only check in V8. Please let me know if there is anything wrong.) 

Yes, pmax is 2 instructions on arm and 1 instruction on x64. min is 1 instruction on arm and many instructions on x64. Relaxed min/max is 1 instruction on both x64 and arm, but has implementation-defined behavior on NaN and +/-0.0.

Since I32x4.dot_i8x16_i7x16_add_s is SDOT on arm64, it will probably go to src/qs8-gemm/MRx4c16-wasmsdot.c.in. src/qs8-gemm/MRx4c16-wasmusdot.c.in may be executed mainly on x64. Shall we keep wasm_f32x4_pmax here in src/qs8-gemm/MRx4c16-wasmusdot.c.in, or replace it with relaxed max (I'm not very confident if it is valid) if only any arm device will use it?